phases/10-llms-from-scratch/21-jamba-hybrid-ssm-transformer/code/main.py

```
@@ -0,0 +1,259 @@
+          f"(attn={b.attn_layers}, mamba={b.mamba_layers}, moe={b.moe_layers})")
+    print(f"  interleave    : {a}:{m}  "
+          f"MoE every {cfg['moe_every']}, {cfg['num_experts']} experts "
+          f"top-{cfg['experts_per_token']}")
+    print(f"  total / active: {fmt_params(b.total_params)} / "
+          f"{fmt_params(b.active_params)}  ({active_frac:.1%})")
+    print(f"  context       : {cfg['max_position_embeddings']:,} tokens")
+    print(f"  KV cache      : {fmt_bytes(b.kv_cache_bytes)} "
+          f"(vs {fmt_bytes(pure)} pure transformer -> {ratio} reduction)")
+    print(f"  SSM state     : {fmt_bytes(b.ssm_state_bytes)} "
+          f"(constant vs seq_len)")
+    print(f"  schedule (Ad=attn+dense, Me=mamba+moe, Md=mamba+dense):")
```

Ruff (0.15.10)

[error] f-string without any placeholders

Remove extraneous `f` prefix

(F541)